Enable filtering of WC REST API responses based on the gla_syncable query parameter

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes part of https://github.com/woocommerce/google-listings-and-ads/issues/2146

This PR allows the filtering of private metadata (metadata prefixed with `_`), excluding metadata with the prefix `_wc_gla`, when a WC REST API request is made using a specific parameter, `gla_syncable`.

Additionally, this PR ensures that only the products the merchant intends to sync are returned. In this case, the required metadata setting is: `_wc_gla_visibility` = `sync-and-show`.

The WC REST API endpoints that need to be filter are the following:

| Endpoint  | Filters |
| ------------- | ------------- |
| `/wp-json/wc/v3/products`  | private metadata + _wc_gla_visibility = sync-and-show  |
| `/wp-json/wc/v3/products/<product_id>`l  | private metadata + _wc_gla_visibility = sync-and-show  |
| `/wp-json/wc/v3/products/<product_id>/variations`  | private metadata  |
| `/wp-json/wc/v3/products/<product_id>/variations/<variation_id>`l  | private metadata  |
| `/wp-json/wc/v3/coupons`l  | private metadata + _wc_gla_visibility = sync-and-show + email restrictions |
| `/wp-json/wc/v3/coupons/<coupon_id>`l  | private metadata + _wc_gla_visibility = sync-and-show + email restrictions  |

Filtering is unnecessary for shipping zone endpoints as they do not contain metadata.

- See here for more context: pcTzPl-1Qd-p2#comment-2663
- Here a list of all endponts: pcTzPl-1Qd-p2#comment-2650

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->



1. Make requests to the endpoints mentioned above and add the `gla_syncable` param, for example: `/wp-json/wc/v3/products?gla_syncable=1`
2. See that the data is filtered correctly.
3. If you remove the query param: `gla_syncable` it should not filter any data. 


### Additional details:

- Variants are special cases since they don't rely on `_wc_gla_visibility` and instead depend on the parent attribute. Hence, if we aim to avoid fetching a variant that the parent product doesn't have the `gla_visibility`, we'd have to check its parent, meaning an additional query. I'm unsure if it's necessary to go to that far, so I'll seek feedback in the project thread.
- Coupons with email restrictions should also be filtered:

![image](https://github.com/woocommerce/google-listings-and-ads/assets/2488994/42dd8ffc-4357-43fc-8abe-6d8e4be4ed88)


<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Add - Filter WC REST API responses for gla_syncable param

Commit:
Update phpunit polyfills to 1.1 for WP 6.4

jorgemd24 committed:
Filter results for WC Proxy

jorgemd24 committed:
WIP tests

jorgemd24 committed:
Rename file

jorgemd24 committed:
Add tests

jorgemd24 committed:
Merge branch 'feature/google-api-project' into add/gla_syncable_filter

jorgemd24 committed:
Fix eslint errors

jorgemd24 committed:
Fix phpcs tests

martynmjones commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,209 @@
+	 *
+	 * @var array
+	 */
+	public static $post_types_to_be_filter = [
```
Should this variable be called `$post_types_to_filter`?

martynmjones commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,209 @@
+	 * @param string $object_type The object type.
+	 */
+	protected function register_object_types_filter( string $object_type ): void {
+		add_filter( 'woocommerce_rest_prepare_' . $object_type . '_object', [ $this, 'check_item_is_syncable' ], 9, 3 );
```
For consistency it would be nice if this was written over multiple lines as the two filters added below are.

martynmjones commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,209 @@
+	 * @return bool
+	 */
+	protected function should_filter_data( WP_REST_Request $request ): bool {
+		return $request->get_param( 'gla_syncable' ) !== null;
```
I appreciate this won't be widely used, however, I'm curious why we're only relying on the presence of the parameter and not paying attention to the value?

If we send a request to both:

- `/wp-json/wc/v3/products?gla_syncable=true`
- `/wp-json/wc/v3/products?gla_syncable=false`

I would expect different results but at the moment we're going to get the same filtered results for both.

jorgemd24 replied:
Initially, my intention was only to check the presence because we include the parameter only for the GL&A app, making the value irrelevant. However, I understand your point that you would expect different results with the above requests so I adjusted the code. I didn't use "true" or "false" because in WPCOM the boolean gets sent as "1" or "0"

martynmjones commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,209 @@
+	/**
+	 * Query items with specific args for example where _wc_gla_visibility is set to sync-and-show.
+	 *
+	 * @param array           $args The query args.
+	 * @param WP_REST_Request $request The request object.
```
The description for `$args` and `$request` are not aligned here.

martynmjones reviewed (commented):
Hey @jorgemd24, thanks for your work adding this filter!

It's working as expected but I've left a few comments asking for clarification and on a few formatting points.

jorgemd24 committed:
Rename post_types_to_filter

jorgemd24 committed:
Tweak formatting

jorgemd24 committed:
Check for the value of gla_syncable

puntope commented:
> Additionally, this PR ensures that only the products the merchant intends to sync are returned. In this case, the required metadata setting is: _wc_gla_visibility = sync-and-show.

I don't see the synced products when I call `/wp-json/wc/v3/products`

Steps:

- Fresh installation
- Do full onboarding 
- Checkout PR
- Call `/wp-json/wc/v3/products?gla_syncable=1` 

Result:

Empty array 

- Go to the product 
- Save
- Call `/wp-json/wc/v3/products?gla_syncable=1` 

Result: 
Product shows


https://github.com/woocommerce/google-listings-and-ads/assets/5908855/cd2f2fce-1135-4f94-b1b3-fdc3ed8eb2a4

When I do Syn all products with Connection Tests it works as well.

puntope commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,215 @@
+	protected function register_object_types_filter( string $object_type ): void {
+		add_filter(
+			'woocommerce_rest_prepare_' . $object_type . '_object',
+			[ $this, 'check_item_is_syncable' ],
```
I would recommend to add a consistent `filter_` name to `check_item_is_syncable` as it's also a filter

puntope commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,215 @@
+	 *
+	 * @return WP_REST_Response The response object updated.
+	 */
+	public function check_item_is_syncable( $response, $item, WP_REST_Request $request ): WP_REST_Response {
```
More than just a check (boolean return) I find this more as a filtering of the response. 

Maybe "filter_response_by_syncable_item" ?? Just an idea, the point is that check asses a true/false statement

puntope commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,215 @@
+			}
+		}
+
+		$data['meta_data'] = array_values( $data['meta_data'] );
```
Why we are just returning the values?

jorgemd24 replied:
Thanks for catching this, I believe it was there because I used `array_filter` to return values and I wanted to reset the indexes. However, it's no longer necessary. I removed here: https://github.com/woocommerce/google-listings-and-ads/pull/2152/commits/5ec640eed1f815a9e83cbfe0a4832b6e9627bce1

puntope commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,215 @@
+	}
+
+	/**
+	 * Filter the metadata of an object.
```
I think it would be more accurate something like:

"Filter the response metadata returning only those prefixed with _wc_gla"

puntope commented on src/Integration/WPCOMProxy.php:
```diff
@@ -0,0 +1,215 @@
+			return $response;
+		}
+
+		$meta_data = $response->get_data()['meta_data'] ?? [];
```
I would  return $response in case `$meta_data === []` because otherwise `$meta->key `will throw a fatal if it happens.

jorgemd24 committed:
Tweaks filter function names and PHPDocs

jorgemd24 committed:
Use the ChannelVisibility constants

jorgemd24 commented:
@martynmjones Thanks for your time reviewing this PR! I addressed all your comments and I replied the question about the `gla_syncable` value.

@puntope 

> I don't see the synced products when I call /wp-json/wc/v3/products

> Steps:

> Fresh installation
> Do full onboarding
Checkout PR
Call /wp-json/wc/v3/products?gla_syncable=1

Is it because the `_wc_gla_visibility` metadata hasn't been added to the products yet? Perhaps the AS takes some time to update the metadata for those products. If that's the case, it's something we could include in the Migration Project.

While investigating the above, I realised that we could use the below constants, so I made changes to the code:

https://github.com/woocommerce/google-listings-and-ads/blob/b84999bfefe337177b41f5bd028c2bdc182042b5/src/Value/ChannelVisibility.php#L17-L18

puntope reviewed (approved):
I tested again this PR and I get the expected results in all the endpoints (products & coupons)

Also, no more inconsistencies from my side in the code. I approve this, so if @martynmjones  doesn't have any issue to address this can be merged.

martynmjones commented:
> Also, no more inconsistencies from my side in the code. I approve this, so if @martynmjones doesn't have any issue to address this can be merged.

No issues for me. Thanks for the changes @jorgemd24
